File: shared/src/commonMain/kotlin/flagent/evaluator/ConstraintEvaluator.kt
package flagent.evaluator

import flagent.evaluator.FlagEvaluator.EvalContext

/**
 * ConstraintEvaluator - evaluates if entity context matches constraint
 * Shared across all platforms (JVM, JS, Native)
 */
class ConstraintEvaluator {
    
    /**
     * Evaluate if all constraints match the evaluation context
     * Returns true if all constraints match, false otherwise
     */
    fun evaluate(
        constraints: List<FlagEvaluator.EvaluableConstraint>,
        context: EvalContext
    ): Boolean {
        if (constraints.isEmpty()) {
            return true
        }
        
        if (context.entityContext == null) {
            return false
        }
        
        return constraints.all { constraint ->
            evaluateConstraint(constraint, context)
        }
    }
    
    /**
     * Evaluate single constraint against context
     */
    private fun evaluateConstraint(
        constraint: FlagEvaluator.EvaluableConstraint,
        context: EvalContext
    ): Boolean {
        val propertyValue = context.entityContext?.get(constraint.property) ?: return false
        val constraintValue = constraint.value
        
        return when (constraint.operator) {
            "EQ" -> equals(propertyValue, constraintValue)
            "NEQ" -> !equals(propertyValue, constraintValue)
            "LT" -> lessThan(propertyValue, constraintValue)
            "LTE" -> lessThanOrEqual(propertyValue, constraintValue)
            "GT" -> greaterThan(propertyValue, constraintValue)
            "GTE" -> greaterThanOrEqual(propertyValue, constraintValue)
            "EREG" -> matchesRegex(propertyValue, constraintValue)
            "NEREG" -> !matchesRegex(propertyValue, constraintValue)
            "IN" -> inList(propertyValue, constraintValue)
            "NOTIN" -> !inList(propertyValue, constraintValue)
            "CONTAINS" -> contains(propertyValue, constraintValue)
            "NOTCONTAINS" -> !contains(propertyValue, constraintValue)
            else -> false
        }
    }
    
    private fun equals(propertyValue: Any, constraintValue: String): Boolean {
        return propertyValue.toString() == constraintValue
    }
    
    private fun lessThan(propertyValue: Any, constraintValue: String): Boolean {
        return compareNumbers(propertyValue, constraintValue) { a, b -> a < b }
    }
    
    private fun lessThanOrEqual(propertyValue: Any, constraintValue: String): Boolean {
        return compareNumbers(propertyValue, constraintValue) { a, b -> a <= b }
    }
    
    private fun greaterThan(propertyValue: Any, constraintValue: String): Boolean {
        return compareNumbers(propertyValue, constraintValue) { a, b -> a > b }
    }
    
    private fun greaterThanOrEqual(propertyValue: Any, constraintValue: String): Boolean {
        return compareNumbers(propertyValue, constraintValue) { a, b -> a >= b }
    }
    
    private fun compareNumbers(
        propertyValue: Any,
        constraintValue: String,
        comparator: (Double, Double) -> Boolean
    ): Boolean {
        return try {
            val propNum = propertyValue.toString().toDouble()
            val constraintNum = constraintValue.toDouble()
            comparator(propNum, constraintNum)
        } catch (e: NumberFormatException) {
            false
        }
    }
    
    private fun matchesRegex(propertyValue: Any, constraintValue: String): Boolean {
        return try {
            val regex = constraintValue.toRegex()
            regex.matches(propertyValue.toString())
        } catch (e: Exception) {
            false
        }
    }
    
    private fun inList(propertyValue: Any, constraintValue: String): Boolean {
        val values = constraintValue.split(",").map { it.trim() }
        return values.contains(propertyValue.toString())
    }
    
    private fun contains(propertyValue: Any, constraintValue: String): Boolean {
        return propertyValue.toString().contains(constraintValue)
    }
}

File: sdk/swift-enhanced/Sources/FlagentEnhanced/Offline/ConstraintEvaluator.swift
import Foundation

/// Evaluates constraints against entity context (EQ, NEQ, IN, etc.). Matches shared ConstraintEvaluator.
struct ConstraintEvaluator {
    func evaluate(constraints: [OfflineLocalConstraint], context: [String: String]) -> Bool {
        if constraints.isEmpty { return true }
        return constraints.allSatisfy { evaluateConstraint($0, context: context) }
    }

    private func evaluateConstraint(_ c: OfflineLocalConstraint, context: [String: String]) -> Bool {
        guard let propertyValue = context[c.property] else { return false }
        let constraintValue = c.value
        switch c.operator_ {
        case "EQ": return equals(propertyValue, constraintValue)
        case "NEQ": return !equals(propertyValue, constraintValue)
        case "LT": return compareNumbers(propertyValue, constraintValue) { $0 < $1 }
        case "LTE": return compareNumbers(propertyValue, constraintValue) { $0 <= $1 }
        case "GT": return compareNumbers(propertyValue, constraintValue) { $0 > $1 }
        case "GTE": return compareNumbers(propertyValue, constraintValue) { $0 >= $1 }
        case "EREG": return matchesRegex(propertyValue, constraintValue)
        case "NEREG": return !matchesRegex(propertyValue, constraintValue)
        case "IN": return inList(propertyValue, constraintValue)
        case "NOTIN": return !inList(propertyValue, constraintValue)
        case "CONTAINS": return propertyValue.contains(constraintValue)
        case "NOTCONTAINS": return !propertyValue.contains(constraintValue)
        default: return false
        }
    }

    private func equals(_ a: String, _ b: String) -> Bool { a == b }

    private func compareNumbers(_ prop: String, _ constraint: String, _ cmp: (Double, Double) -> Bool) -> Bool {
        guard let a = Double(prop), let b = Double(constraint) else { return false }
        return cmp(a, b)
    }

    private func matchesRegex(_ prop: String, _ pattern: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
        let range = NSRange(prop.startIndex..., in: prop)
        return regex.firstMatch(in: prop, range: range) != nil
    }

    private func inList(_ prop: String, _ list: String) -> Bool {
        let values = list.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
        return values.contains(prop)
    }
}
